src/linker/Linker/DependencyInfo.cs

```diff
@@ -0,0 +1,142 @@
+		CustomAttribute,
+		ParameterAttribute,
+		ReturnTypeAttribute,
+		GenericParameterCustomAttribute,
+		GenericParameterConstraintCustomAttribute,
+
+	// Dependencies of custom attributes
+		AttributeConstructor, // attribute -> its ctor
+		// used for security attributes, where we mark the type/properties directly
+		AttributeType, // attribute -> attribute type
+		AttributeProperty, // attribute -> attribute ptroperty
+		CustomAttributeArgumentType,
+		CustomAttributeArgumentValue,
+		CustomAttributeField,
+
+	// Tracking cctors
+		TriggersCctorThroughFieldAccess, // field-accessing method -> cctor of field's declaring type
+		TriggersCctorForCalledMethod, // caller method -> callee method type cctor
+		DeclaringTypeOfCalledMethod, // called method -> its declaring type (used to track when a cctor is triggered by a method call)
+		CctorForType, // type -> cctor of type
+		CctorForField, // field -> cctor of field's declaring type
+
+	// Tracking instantiations
+		InstantiatedByCtor, // ctor -> type it instantiates			
+		OverrideOnInstantiatedType, // instantiated type -> override method on the type
+
+	// Linker-specific behavior (preservation hints, patterns, user inputs, linker outputs, etc.)
+		PreservedDependency, // PreserveDependency attribute -> member
+		AccessedViaReflection, // method -> detected member accessed via reflection from that method
+		PreservedMethod, // type/method -> preserved method (explicitly preserved in Annotations by XML or other steps)
+		TypePreserve, // type -> field/method preserved for the type (explicitly set in Annotations by XML or other steps)
+		DisablePrivateReflection, // type/method -> DisablePrivateReflection attribute added by linkerf
+
+	// Built-in knowledge of special runtime/diagnostic subsystems
+		// XmlSchemaProvider, DebuggerDisplay, DebuggerTypeProxy, SoapHeader, TypeDescriptionProvider
+		ReferencedBySpecialAttribute, // attribute -> referenced members
+		KeptForSpecialAttribute, // attribute -> kept members (used when the members are not explicitly referenced)
+		SerializationMethodForType, // type -> method required for serialization
+		EventSourceProviderField, // EventSource derived type -> fields on nested Keywords/Tasks/Opcodes provider classes
+		MethodForSpecialType, // type -> methods kept (currently used for MulticastDelegate)
+
+	// Linker internals, requirements for certain optimizations
+		UnreachableBodyRequirement, // method -> well-known type required for unreachable bodies optimization
+		DisablePrivateReflectionRequirement, // null -> DisablePrivateReflectionAttribute type/methods (note that no specific source is reported)
+		AlreadyMarked, // null -> member that has already been marked for a particular reason (used to propagate reasons internally, not reported)
+
+	// For use when we don't care about tracking a particular dependency
+		Unspecified, // currently unused
+	}
+
+	readonly public struct DependencyInfo : IEquatable<DependencyInfo> {
```

Comment:
Do you have any suggestions on how we could extend this mechanism with additional dependency information that is unity specific?

Modifying the DependencyKind enum in our fork is bound to become a maintenance nightmare.

Reply:
I think it should be fine to include your values here as well

Reply:
> I think it should be fine to include your values here as well
I'd rather not do that.  That means every time we add a new value to our linker we'd have to remember to take it upstream.

@vitek-karas I don't entirely follow what that would look like.  Here is an example of one thing we do.  Can you show me what that would look like with your proposal?

https://gist.github.com/mrvoorhe/4711e93688968f4a8a13dda2344c4727

Reply:
For example:
``` C#
class MonoBehaviourAnnotation
{
    public bool IsMainScript { get; set; }
}

        protected void ProcessType(TypeDefinition typeDefinition)
        {
            using (Tracer.PushScoped(typeDefinition))
            {
                 // Mark all types derived from MonoBehaviour
                var dependencyInfo = new DependencyInfo(DependencyKind.Unspecified, "Built in rule for MonoBehaviour");
                dependencyInfo.AddAnnotation(new MonoBehaviourAnnotation() { IsMainScript = false });
                _context.Annotations.Mark(typeDefinition, dependencyInfo);
            }
        }

        // Implementation of IDependencyResolver
        override void RecordDependency(object target, DependencyInfo reason, bool marked)
        {
            if (reason.GetAnnotation<MonoBehaviourAnnotation>()?.IsMainScript)
                 Console.WriteLine($"{target} is the main script");
        }

Reply:
What if I added an optional third `object` field as you suggest, but made it public? This way the DependencyInfo struct can remain immutable (the extra info would be passed to a new ctor), and extenders can use that field however they like - including by defining a new type with generic a AddAnnotation<T> method?

Reply:
I guess it would work, but it's not as "nice" :-)
The advantage of the Type-as-a-key approach and have it directly on the info object is that anything in the system can add annotation without any other part of the system being affected by it (unless they explicitly cooperate). If we add a public object then we would have to make sure that everything which uses it knows how to cooperate with anything else which uses it. As an example if Unity would use it to store some additional info, and the MarkStep itself would want to use it store additional information about the dependency (field for example), then these two would somehow have to cooperate since they would have to share that one field to store the value in.

What are the advantages of keeping the struct readonly? Does it make it significantly easier for JIT to optimize or something else?

Reply:
I mainly used readonly to express the intention that DependencyInfo is a simple data type that only carries information about why a cecil object was kept. There might be some perf advantages like preventing copies for in parameters - I don't know enough about the JIT to say whether it can better optimize readonly structs.

I actually see readonly as enforcing that we don't need to worry about the kind of cooperation you mention. Whoever creates a new DependencyInfo must fully populate it, and shouldn't rely on any other part of the marking logic to change it by adding or removing information to/from it. It's a simpler conceptual model, and I usually like to keep things simple unless there's a good reason. :) How likely do you think it is that we'll need to support the kind of interaction where both Unity and MarkStep add information to the same dependency reason?

Another line of reasoning is that we already have AnnotationStore as the core mutable database of information we are tracking about the app's dependencies, and adding "annotations" to dependencies seems slightly redundant (there are definitely things you could do with annotations on DependencyInfo that couldn't be done with the current Annotations... I just can't think of a good example where we would want to).

It also occurs to me that we could reuse the `Source` field, maybe with a `DependencyKind.Custom`, to reduce the memory footprint.

Reply:
Sounds like there are pros & cons to either approach. 

A 3rd field that is reserved for us would be OK.  A 3rd field that is used by both monolinker and unitylinker would be painful to maintain as @vitek-karas  points out.

Do what you can to put something in place.  Whatever you think is best.  When we pull this change in I'll try it out and see how it goes and we can refine as necessary.

Reply:
@sbomer you have a good point about the AnnotationStore - we could add that if needed.
I'm fine with whatever you decide here.

Reply:
I'll add a `DependencyKind.Custom` with a `Source` reserved for use by extenders. If we find a case where that's too restrictive, we can change this.